File: Chan_Jeliazkov_2009/kalman_filter.R
#------------------------------------------------------------------------------
# R Code for a fast Kalman Filter based on 
# Chan, J.C.C. and Jeliazkov, I. (2009). Efficient Simulation and
# Integrated Likelihood Estimation in State Space Models,
# International Journal of Mathematical Modelling and Numerical
# Optimisation, 1, 101-120.
#------------------------------------------------------------------------------

library(Matrix)
library(SuppDists)
library(RcppZiggurat)
library(Rcpp)


# TVP - VAR #---------------------------------------------------------------
# VAR:
# y[t] = mu[t] + Gamma[t] * y[t-1] + epsilon[t],   epsilon[t] ~ N(0, Omega_11)
# 
# State Space Form
# y[t] = X[t] * beta[t] + epsilon[t]   epsilon[t] ~ N(0, Omega_11)
# beta[t]=beta[t-1] + nu[t]     nu[t] ~ N(0, Omega_22)
#
# beta[t] = vec(mu[t], Gamma[t]')'
# 

rm(list = ls())


set.seed(123)

nsim <- 100
nburn <- 0
total_runs <- nsim + nburn

data <- read.csv("./papers/Chan_Jeliazkov_2009/USdata.csv", header = FALSE)
colnames(data) <- c("gpd_growth", "unemp", "inter", "inf")
y <- ts(data = data,
        start = c(1948, 1),
        freq = 4)

#y <- c(log(y)) * 100
y <- as.matrix(y)
colnames(y) <- NULL
y0 <- y[1:3, ]
#y <- y[-c(1:3), ]

Y <- matrix(c(t(y[-c(1:3),])))

tt <- nrow(y)
nn <- ncol(y)
TT <- nrow(Y)/nn
qq <- nn*(nn+1)
TTqq <- TT*qq
# priors #---------------------------------------------------------------------

# Omega_11
nu1 <- nn + 3 # nn?
S1 <- diag(nn)

# Omega_22
DD <- 5 * diag(qq)
nu2 <- rep(6, qq)
S2 <- rep(0.01, qq)


# initial values #-------------------------------------------------------------

# Omega_11
Omega11 <- Matrix(cov(y))
Omega11_inv <- Matrix(solve(Omega11, diag(ncol(Omega11))))

# H
diags_H <- list(rep(1, TT*qq), rep(-1, (TT - 1)*qq))
H <- bandSparse(n = TT*qq,
               k = c(0, -(qq+1)),
               diag = diags_H,
               symm = FALSE
               )

# S
DD_inv <- solve(DD)
Omega22 <- 0.01*diag(qq) # initial values for variance of betas?
Omega22_inv <- solve(Omega22)
S <- bdiag(replicate((TT-1), Omega22, simplify = F))
S <- bdiag(DD, S)  

S_inv <- bdiag(replicate((TT-1), Omega22_inv, simplify = F))
S_inv <-  bdiag(DD_inv, S_inv)  

# bigG = X
#G <- bdiag(replicate(qq, matrix(c(1, y[3:(TT-1),]), ncol=nn+1), simplify = F))

G_list <- vector("list", 245)
for (i in 3:247){
    G_list[[i-2]] <- kronecker(diag(4), matrix(c(1, y[i,]), ncol=nn+1))
}
G <- bdiag(G_list)

new_nu2 <- (nu2+(TT-1))/2

# store
store_beta <- array(NA, dim=c(nsim, TT*qq))
store_Omega11 <- array(NA, dim=c(nn, nn, nsim))
store_Omega22 <- array(NA, dim=c(nsim, qq))
now <- Sys.time()
# Gibbs Sampler -----------------------------------------------------------
for (ii in 1:total_runs) {

# beta --------------------------------------------------------------------
    S_inv <- kronecker(Diagonal(TT), Omega22_inv)
    S_inv[1:qq, 1:qq] <- DD_inv 
    K <- Matrix::forceSymmetric(crossprod(H, S_inv) %*% H)
    G_Omega11_inv <- crossprod(G, kronecker(diag(TT), Omega11_inv))
    G_Omega11_inv_G <-  Matrix::forceSymmetric(G_Omega11_inv %*% G)
    P <- K + G_Omega11_inv_G
    L <- chol(P)
    beta_hat <- Matrix(backsolve(L, forwardsolve(L, G_Omega11_inv  %*% Y, upper.tri = FALSE),
                          upper.tri = TRUE, transpose = TRUE))
    beta <- beta_hat + solve(L, zrnorm(n = length(beta_hat)))
    
# Omega_11 ----------------------------------------------------------------
    e1 <- matrix(Y-G%*%beta, nrow=nn)
    new_S1 <- S1 + tcrossprod(e1) 
    Omega11_inv <- rWishart(n = 1, df = nu1 + TT, Sigma =  solve(new_S1, diag(nn)))[,,1]
    Omega11 <- solve(Omega11, diag(nn))

# Omega_22 |---------------------------------------------------------------
    e2 <- matrix(H %*% beta, ncol=qq)
    new_S2 <- (S2 + colSums(e2[-1,]**2))/2
    diag(Omega22) <- 1/rgamma(n=20,
                 shape = new_nu2,
                 scale = 1/new_S2
                     )

# store |------------------------------------------------------------------
    if (ii > nburn) {
        run <- ii - nburn
        store_beta[run, ] <- beta@x
        store_Omega11[ , , run] <- matrix(Omega11@x, nrow=nn)
        store_Omega22[run, ] <- diag(Omega22)
    }
}
print(Sys.time() - now)


# Results #--------------------------------------------------------------------
beta_hat <- colMeans(store_beta)
Omega11_hat <- apply(store_Omega11, MARGIN = c(1,2), FUN = mean)
Omega22_hat <- colMeans(store_Omega22)

y_hat <- matrix(beta_hat, nrow=qq)
XX <- array(NA, c(4,5,245))
for (i in 1:245){
    XX[,,i] <- matrix(y_hat[,i], nrow=nn)
}


dimnames(XX) <- list(NULL, c("mu", "gdp", "une", "int", "inf"), NULL)
gdp <- t(XX[1,,])
une <- t(XX[2,,])
int <- t(XX[3,,])
inf <- t(XX[4,,])
#reshape2::melt(gdp)

ggplot2::ggplot(reshape2::melt(inf),
    ggplot2::aes(x = Var1, y = value, col=Var2)) +
    ggplot2::geom_line()+
    ggplot2::ylim(c(-0.4,0.4))

File: Chan_Jeliazkov_2009/kalman_filter.jl
#------------------------------------------------------------------------------
# Julia Code for a fast Kalman Filter based on
# Chan, J.C.C. and Jeliazkov, I. (2009). Efficient Simulation and
# Integrated Likelihood Estimation in State Space Models,
# International Journal of Mathematical Modelling and Numerical
# Optimisation, 1, 101-120.
#------------------------------------------------------------------------------
using LinearAlgebra
using CSV
using SparseArrays
using Distributions
using Random
using Plots

# 1) TVP-VAR
function SUR(X)
    r, c = size(X);
    idi = reshape((kron((1:r), ones(c, 1) )), r*c);
    idj = collect(1:r*c);
    X_SUR = sparse(idi, idj, reshape(X',r*c));
    return X_SUR
end

function gibbs_sampler(y::AbstractArray{Float64, 2}, nsim::Int64)

    epss = eps()^0.4

    y0 = y[1:2, :];
    Y = y[4:end, :];

    t = size(y, 1);
    n = size(y, 2);
    T = size(Y, 1);
    q = n*(n+1);
    Tq = T*q;

    Y = reshape(Y',:,1);

    # priors #-----------------------------------------------------------------
    # Omega_11
    nu01 = n + 3.;
    S01 = 1.0I(n);

    # Omega_22
    DD = 5. * sparse(1.0I, q, q);
    DD_inv = 1.0/5 * sparse(1.0I, q, q);
    nu02 = 6. * ones(q);
    S02 = 0.01 * ones(q);

    # initial values #---------------------------------------------------------
    new_nu01 = nu01 + T;
    new_nu02 = nu02/2 .+ (T-1)/2;

    # Omega11
    Omega11 = cov(y);
    Omega11_inv = inv(Omega11);

    # H
    H1 = sparse(1.0I, Tq, Tq);
    H2 = [[zeros(q, Tq-q); sparse(1.0I, Tq-q, Tq-q)] zeros(Tq, q)];
    H = (H1 - H2);
    Ht = sparse(H');

    # S
    Omega22 = 0.01 * sparse(1.0I, q, q);
    Omega22_inv = 10.0 * sparse(1.0I, q, q);
    S = blockdiag(DD, kron(sparse(1.0I, T-1, T-1), Omega22));
    S_inv = blockdiag(DD_inv, kron(sparse(1.0I, T-1, T-1), Omega22_inv));

    # G
    bigG = SUR([ones(T*n,1) kron(y[3:end-1, :], ones(n))]);
    bigGt = sparse(bigG');

    # initialize for storeage
    store_beta = zeros(Tq, nsim);
    store_Omega11 = zeros(n, n,  nsim);
    store_Omega22 = zeros(q, nsim);

    for isim in 1:nsim

        S_inv = blockdiag(DD_inv, kron(sparse(I,T-1,T-1), Omega22_inv));
        K = Ht * S_inv * H;

        GtOmega11inv = bigGt * kron(sparse(1.0I, T, T), Omega11_inv);
        GtOmega11invG = GtOmega11inv*bigG;
        GtOmega11invY = GtOmega11inv * Y;

        P = K + GtOmega11invG;
        P = (P+sparse(P'))./2

        tmp = cholesky(P, perm = 1:Tq, check = false);
        if issuccess(tmp)
            L = sparse(tmp.L);
        else
            tmp = ldlt(P, perm = 1:Tq, check = false);
            if issuccess(tmp)
                LD = sparse(tmp.LD);
                e = diag(LD);
                e[findall(x -> x <= epss, e)] .= 0;
                L = LD - Diagonal(LD) + I(Tq);
                D = sparse(Diagonal(e));
                L*D*sparse(L') ≈    P
                L = L*sparse(Diagonal(e .^0.5))
            end

        end

        beta_hat = sparse(L')\(L\GtOmega11invY);
        beta = beta_hat + sparse(L') \ rand(Normal(),Tq);

        # Omega11
        e1 = reshape(Y - bigG * beta, n,:);
        new_S01 = S01 + e1*e1';
        Omega11 = rand(InverseWishart(new_nu01, new_S01));
        Omega11_inv = sparse(Symmetric(Omega11\I(n)));

        # Omega22
        e2 = reshape(H * beta, q, T)';
        new_S02 = (S02 + sum(e2[2:end,:].^2, dims=1)[:])/2;
        for i in 1:q
            Omega22[i,i] = rand(InverseGamma(new_nu02[i], new_S02[i]));
        end
        Omega22_inv = Omega22\sparse(1.0I, q, q);

        # store
        store_beta[:, isim] = beta;
        store_Omega11[:,:, isim] = Omega11;
        store_Omega22[:, isim] = diag(Omega22);
    end

    return store_beta, store_Omega11, store_Omega22
end

# Kalman Filter
data_raw = CSV.read("./Chan_Jeliazkov_2009/USdata.csv", header = 0);
y = Matrix(data_raw);


@time store_beta, store_Omega11, store_Omega22  = gibbs_sampler(y, 10);
@time store_beta, store_Omega11, store_Omega22  = gibbs_sampler(y, Int(2.5e4));

beta_hat = mean(store_beta, dims = 2)';
Omega11_hat = mean(store_Omega11, dims = 3)[:,:,1];
Omega22_hat = mean(store_Omega22, dims = 2)';

beta = reshape(beta_hat, 4, 5, :);

l = @layout [a b ; c d];
p1 = plot(1:245, beta[1,:,:]', legend=false);
p2 = plot(1:245, beta[2,:,:]', legend=false);
p3 = plot(1:245, beta[3,:,:]', legend=false);
p4 = plot(1:245, beta[4,:,:]', legend=false);
plot(p1, p2, p3, p4, layout = l)

savefig( "./Chan_Jeliazkov_2009/fg1")

# 2) Dynamic Factor Model
